File: SeleniumEnhanced.py
from selenium import webdriver
import getpass
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from selenium.common.exceptions import *
import time
from pandas import read_excel
import pandas as pd
from selenium.webdriver.chrome.options import Options


class SeleniumEnhanced:

    def __init__(self):

        self.userHMF = getpass.getuser()
        self.download_dir = "C:/Users/" + self.userHMF + "/Downloads"
        self.options = Options()  # add options to skip "Save as" prompts and set save directory
        self.options.add_experimental_option("prefs", {
            "download.default_directory": self.download_dir,
            "download.prompt_for_download": False,
            "download.directory_upgrade": True,
            "safebrowsing.enabled": True
        })
        # error when runnning in command prompt, will fix later
        #self.driver = webdriver.Chrome("C:/Users/" + getpass.getuser() + "/Python_Files/chromedriver.exe",
         #                              options=self.options)

        # delete next line when above line is fixed
        self.driver = webdriver.Chrome("C:/Users/" + getpass.getuser() + "/ImportFiles/chromedriver.exe")
        self.driver.maximize_window()
        self.i = 0
        self.MaxTries = 20
        self.SecondsToWait = 2
        self.FatalError = False
        self.ExitOnError = True
        self.ShowMessages = True

    # doesn't work correctly, will figure out later
    def show_messages(self, TrueOrFalse):
        self.ShowMessages = bool(TrueOrFalse)

    def click(self, xpath, catchoutput):
        self.i = 0
        while self.i <= (self.MaxTries-1):
            try:
                self.driver.find_element_by_xpath(xpath).click()
                self.driver.implicitly_wait(10)
                self.i = self.MaxTries
            except Exception:
                if self.ShowMessages:
                    print(catchoutput + ": try " + str(self.i + 1) + " out of " + str(self.MaxTries))
                self.i += 1
                time.sleep(self.SecondsToWait)
                if self.i == self.MaxTries:
                    self.FatalError = True
                    if self.ExitOnError:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getText")
                        self.driver.close()
                        self.driver.quit()
                        #exit()
                    else:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getText")

    def sendKeys(self, xpath, catchoutput, keystosend):
        self.i = 0
        while self.i <= (self.MaxTries-1):
            try:
                self.driver.find_element_by_xpath(str(xpath)).send_keys(str(keystosend))
                self.driver.implicitly_wait(10)
                self.i = self.MaxTries
            except Exception:
                if self.ShowMessages:
                    print(catchoutput + ": try " + str(self.i + 1) + " out of " + str(self.MaxTries))
                self.i += 1
                time.sleep(self.SecondsToWait)
                if self.i == (self.MaxTries):
                    self.FatalError = True
                    if self.ExitOnError:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:sendKeys")
                        self.driver.close()
                        self.driver.quit()
                        #exit()
                    else:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:sendKeys")

    def clear(self, xpath, catchoutput):
        self.i = 0
        while self.i <= (self.MaxTries-1):
            try:
                self.driver.find_element_by_xpath(xpath).clear()
                self.driver.implicitly_wait(10)
                self.i = self.MaxTries
            except Exception:
                if self.ShowMessages:
                    print(catchoutput + ": try " + str(self.i + 1) + " out of " + str(self.MaxTries))
                self.i += 1
                time.sleep(self.SecondsToWait)
                if self.i == self.MaxTries:
                    self.FatalError = True
                    if self.ExitOnError:
                        print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:clear")
                        self.driver.close()
                        self.driver.quit()
                        #exit()
                    else:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:clear")

    def getText(self, xpath, catchoutput):
        self.i = 0
        txt = None
        while self.i <= (self.MaxTries-1):
            try:
                txt = self.driver.find_element_by_xpath(xpath).text
                self.driver.implicitly_wait(10)
                self.i = self.MaxTries
            except Exception:
                if self.ShowMessages:
                    print(catchoutput + ": try " + str(self.i + 1) + " out of " + str(self.MaxTries))
                self.i += 1
                time.sleep(self.SecondsToWait)
                if self.i == self.MaxTries:
                    self.FatalError = True
                    if self.ExitOnError:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getText")
                        self.driver.close()
                        self.driver.quit()
                        #exit()
                    else:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getText")
        return txt

    def getValue(self, xpath, catchoutput):
        self.i = 0
        txt = None
        while self.i <= (self.MaxTries-1):
            try:
                txt = self.driver.find_element_by_xpath(xpath).get_attribute("value")
                self.driver.implicitly_wait(10)
                self.i = self.MaxTries
            except Exception:
                if self.ShowMessages:
                    print(catchoutput + ": try " + str(self.i + 1) + " out of " + str(self.MaxTries))
                self.i += 1
                time.sleep(self.SecondsToWait)
                if self.i == self.MaxTries:
                    self.FatalError = True
                    if self.ExitOnError:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getValue")
                        self.driver.close()
                        self.driver.quit()
                        #exit()
                    else:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getValue")
        return txt

    # will work on later, trying to compress all functions into one method/function
    def getFunction(self, fuction, xpath, catchoutput):
        self.i = 0
        txt = None
        while self.i <= (self.MaxTries-1):
            try:
                txt = self.driver.find_element_by_xpath(xpath).text
                self.driver.implicitly_wait(10)
                self.i = self.MaxTries
            except Exception:
                if self.ShowMessages:
                    print(catchoutput + ": try " + str(self.i + 1) + " out of " + str(self.MaxTries))
                self.i += 1
                time.sleep(self.SecondsToWait)
                if self.i == self.MaxTries:
                    self.FatalError = True
                    if self.ExitOnError:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getText")
                        self.driver.close()
                        self.driver.quit()
                        #exit()
                    else:
                        if self.ShowMessages:
                            print("STOPPING PROCESS - FATAL ERROR: " + catchoutput + " | " + "Action:getText")
        return txt

File: SeleniumEnhanced.java
package globalclasses;


import java.util.concurrent.TimeUnit;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.ie.InternetExplorerDriver;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.Select;

public class SeleniumEnhanced {
	
	ChromeOptions options = new ChromeOptions();
		
	private String userHMF = System.getProperty("user.name");
	public WebDriver driver = new ChromeDriver();
    public Actions action = new Actions(driver);
    
    	
	private String co; //Catch output
	private String xp; //xPath
	private int mt = 5; // MaxTries
	private String kts; // Keys to Send
	private String gt; //return values for getText method/function
	private String an; //Attribute Name;
	private String on; //Option name
	private boolean FatalError = false; 
	public boolean eoe = true;
	
	//alter variables here
	private int SecondsToWait = 10;
	
	
	public void setErrorExit(boolean ExitOnError){
		eoe = ExitOnError;
	}
	
	public void setMaxTries(int MaxTries){
		mt = MaxTries;
	}
	
	public void setChromeDriver(String Location){
		
		System.setProperty("webdriver.chrome.driver", Location);
		
	}
	
	public void setIEDriver(String Location){
		
		driver = new InternetExplorerDriver();
		System.setProperty("webdriver.chrome.driver", "C:\\Users\\" + userHMF + "\\Java_Files\\IEDriverServer.exe");
		
	}
	
	public String getText(String XPath, String CatchOutput) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;
		

		for(int i=0; i < (mt);){

			try{  
				gt = driver.findElement(By.xpath(xp)).getText(); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					FatalError = true;
					System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:getText");
					driver.close();
					driver.quit();
					System.exit(0); 
				}
				i++;
							}
		}

		return gt; 
	}
	
	public String getText(String XPath, String CatchOutput, boolean ExitOnError) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;
		eoe = ExitOnError;

		for(int i=0; i < (mt);){

			try{  
				gt = driver.findElement(By.xpath(xp)).getText(); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);

				if(i == (mt-1)){
					FatalError = true;
					if (eoe = ExitOnError){
						System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:getText");
						driver.close();
						driver.quit();
						System.exit(0); 
					}
					else { System.out.println("FATAL ERROR: " + CatchOutput  + " | " + "Action:getText");}
				}
			}
			i++;

		}

		return gt; 
	}
	
	
	
	public String getAttribute(String XPath, String CatchOutput, String AttributeName) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;
		an = AttributeName;
		eoe = true;

		for(int i=0; i < (mt);){

			try{  
				gt = driver.findElement(By.xpath(xp)).getAttribute(an); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					FatalError = true;
					System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:getAttribute");
					driver.close();
					driver.quit();
					System.exit(0); 
				}
				i++;


			}
		}
		return gt;

	}
	
	//Currently having issues with this method
	public String selectDropDownOption(String XPath, String CatchOutput, String OptionName) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;
		on = OptionName;
		eoe = true;

		for(int i=0; i < (mt);){

			try{  
				Select option = new Select(driver.findElement(By.xpath(xp))); 
				option.selectByVisibleText(on);				
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					FatalError = true;
					System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:getAttribute");
					driver.close();
					driver.quit();
					System.exit(0); 
				}
				i++;


			}
		}
		return gt;

	}
	
	
	public void click(String XPath, String CatchOutput) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;		

		for(int i=0; i < (mt);){

			try{  
				driver.findElement(By.xpath(xp)).click(); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					if (eoe == true){
						FatalError = true;
						System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:click");
						driver.close();
						driver.quit();
						System.exit(0);
					}
					else { 
						System.out.println("FATAL ERROR: " + CatchOutput + " | " + "Action:click");
						//					throw new Exception(CatchOutput + " | " + "Action:click") ; 
					}

				}
				if(i == mt/2){
					driver.navigate().refresh();
				}
				i++;


			}
		}
	}
	
	public void click(String XPath, String CatchOutput, boolean ExitOnError) throws Exception{

		xp = XPath;
		co = CatchOutput;

		for(int i=0; i < (mt);){

			try{  
				driver.findElement(By.xpath(xp)).click(); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					if (eoe = ExitOnError){
						FatalError = true;
						System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:click");
						driver.close();
						driver.quit();
						System.exit(0);
					}
					else { 
						System.out.println("FATAL ERROR: " + CatchOutput + " | " + "Action:click");
						throw new Exception(CatchOutput + " | " + "Action:click") ; 
					}
				}
				if(i == mt/2){
					driver.navigate().refresh();
				}
				i++;


			}
		}
	}

	public void clear(String XPath, String CatchOutput) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;
		eoe = true;

		for(int i=0; i < (mt);){

			try{  
				driver.findElement(By.xpath(xp)).clear(); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					FatalError = true;
					System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:clear");
					driver.close();
					driver.quit();
					System.exit(0); 
				}
				i++;


			}
		}


	}

	
	public void sendKeys(String XPath, String CatchOutput, String KeysToSend) throws InterruptedException{

		xp = XPath;
		co = CatchOutput;
		kts = KeysToSend; 
		eoe = true;

		for(int i=0; i < (mt);){

			try{  
				driver.findElement(By.xpath(xp)).sendKeys(kts); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);
				if(i == (mt-1)){
					FatalError = true;
					System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:sendKeys");
					driver.close();
					driver.quit();
					System.exit(0); 
				}

			}
			//handles internet interruptions. Once variable i meet half of max attempts the page will be reloaded.
			if(i == mt/2){
				driver.navigate().refresh();
			}
			i++;
		}
	}

	
	public void sendKeys(String XPath, String CatchOutput, String KeysToSend, boolean ExitOnError) throws Exception{

		xp = XPath;
		co = CatchOutput;
		kts = KeysToSend; 

		for(int i=0; i < (mt);){

			try{  
				driver.findElement(By.xpath(xp)).sendKeys(kts); 
				i = mt;
			}
			catch (Exception e){
				System.out.println(co + ": try " + (i+1) + " out of " + mt);
				TimeUnit.SECONDS.sleep((long) SecondsToWait);

				if(i == (mt-1)){
					FatalError = true;
					if (eoe = ExitOnError){
						System.out.println("STOPPING PROCESS - FATAL ERROR: " + CatchOutput + " | " + "Action:sendKeys");
						driver.close();
						driver.quit();
						System.exit(0); 
					}
					else { 
						System.out.println("FATAL ERROR: " + CatchOutput + " | " + "Action:sendKeys");
						throw new Exception(CatchOutput + " | " + "Action:sendKeys") ; 
						}
				}
			}
			if(i == mt/2){
				driver.navigate().refresh();
			}
			i++;
		}

	}
	
	

}
